taichi/backends/opengl/opengl_api.cpp

```diff
@@ -525,23 +572,33 @@ struct CompiledProgram::Impl {
   }
 
   void launch(Context &ctx, GLSLLauncher *launcher) const {
-    GLBufferTable &bufs = launcher->impl->user_bufs;
+    launcher->impl->gl_allocator.new_frame();
```

Comment:
nit: could you break this function into smaller pieces?

Reply:
It's kind of hard, I removed the special case handling for the one external ptr case (the logic are unified), but overall there's just a lot of code here to sync buffers in and out